pkg/api/validation/validation.go

```diff
@@ -1278,6 +1291,18 @@ func ValidateSecret(secret *api.Secret) errs.ValidationErrorList {
 		}
 	case api.SecretTypeOpaque, "":
 		// no-op
+	case api.SecretTypeDockercfg:
+		dockercfgBytes, exists := secret.Data[api.DockerConfigKey]
+		if !exists {
+			allErrs = append(allErrs, errs.NewFieldRequired(fmt.Sprintf("data[%s]", api.DockerConfigKey)))
+			break
+		}
+
+		dockercfg := &credentialprovider.DockerConfig{}
+		if err := json.Unmarshal(dockercfgBytes, dockercfg); err != nil {
```

Comment:
Currently there's no good way to test a dockercfg that I am aware of
without actually pulling an image.  We've discussed the need for a way to
do a dry-pull or credential-verify operation on a registry.  I would love
it if someone had a good way to test the credentials.

I agree with @lavalamp that making a validation on the format makes it part
of the API -- I'm not sure we want that.  Do we have any precedent for
validating the content of secrets at a deeper level than which keys are
present?  @liggitt?

cc @ncdc

On Sat, May 16, 2015 at 12:44 PM, David Eads <notifications@github.com>
wrote:

> In pkg/api/validation/validation.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7966#discussion_r30462486>
> :
>
> > @@ -1278,6 +1291,18 @@ func ValidateSecret(secret *api.Secret) errs.ValidationErrorList {
> >  		}
> >  	case api.SecretTypeOpaque, "":
> >  		// no-op
> > +	case api.SecretTypeDockercfg:
> > +		dockercfgBytes, exists := secret.Data[api.DockerConfigKey]
> > +		if !exists {
> > +			allErrs = append(allErrs, errs.NewFieldRequired(fmt.Sprintf("data[%s]", api.DockerConfigKey)))
> > +			break
> > +		}
> > +
> > +		dockercfg := &credentialprovider.DockerConfig{}
> > +		if err := json.Unmarshal(dockercfgBytes, dockercfg); err != nil {
>
>  If it's because kubelet is going to need to parse & understand the
> .dockercfg file, then effectively this is part of the API and this struct
> definition needs to be versioned etc.
>
> It does need to parse in order to be used, but the content of secret data
> is actually a format owned by docker, so if we forced it to have a version
> and type then it would no longer be in the correct format.
>
> Validating that it parses doesn't really give any assurance that it'll
> actually work with docker.
>
> It is necessary, but not sufficient. However, I don't feel strongly enough
> about it to force its inclusion. If you're unconvinced, I'll remove it on
> Monday.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7966/files#r30462486>
> .
>